### PR TITLE
PS-8446 Document LDAP plugin changes in 8.0.30

### PR DESCRIPTION
**Pull request by patrickbirch**
M  docs/security/simple-ldap-variables.md
M  docs/security/simple-ldap.md

### CODE REVIEW

File: docs/security/simple-ldap.md [hunk 1]
```diff
@@ -1,42 +1,124 @@
...
+
+Add the following statements to your `my.cnf` file:
+
+```text
+[mysqld]
+plugin-load-add=authentication_ldap_simple.so
+authentication_ldap_simple_server_host=127.0.0.1
+authentication_ldap_simple_bind_base_dn='dc=percona, dc=com'
+plugin-load-add=authentication_ldap_sasl.so
+authentication_ldap_sasl_server_host=127.0.0.1
```
Comment (dutow) [inline docs/security/simple-ldap.md]: this example mixes simple and sasl plugins

File: docs/security/simple-ldap.md [hunk 2]
```diff
@@ -45,54 +127,117 @@ plugin in the `CREATE USER` statement or the `ALTER USER` statement.:
...
+mysql> CREATE USER 'ldapuser'@'localhost' IDENTIFIED WITH authentication_ldap_simple BY 'uid=ldapuser,ou=testusers,dc=percona,dc=com';
+```
+
+The authentication string does not include the LDAP password. This password must be provided by the client user when they connect.
+
+```sql
+mysql> mysql --user=ldapuser --password --enable-cleartext-plugin
+```
+
+The user enters the `ldapuser` password. The client sends the password as cleartext, which is necessary when using a server-side LDAP library without SASL. This requirement is a security problem in some environments. The following actions may minimize the risk:
```
Comment (dutow) [inline docs/security/simple-ldap.md]: "This requirement is a security problem in some environments." - I'm not sure if I would add this, the SASL setup has its own security issues.

From a security perspective, ldap_simple with an SSL connection is more secure than the SASL plugin (with our without SSL), and the SASL plugin is more secure than ldap_simple without SSL.

File: docs/security/simple-ldap-variables.md [hunk 3]
```diff
@@ -1,4 +1,28 @@
...
+
+The installation adds the following variables:
+
+| Variable name | Description |
+|---|---|
+| authentication_ldap_simple_bind_base_dn | Base distinguished name |
+| authentication_ldap_simple_bind_root_dn | Root distinguished name |
+| authentication_ldap_simple_bind_root_dn_pwd | Password for the root distinguished name |
+| authentication_ldap_simple_ca_path | Absolute path of the certificate authority |
+| authentication_ldap_simple_group_role_mapping | LDAP group name |
```
Comment (dutow) [inline docs/security/simple-ldap-variables.md]: No, this variable contains a list of ldap group name - mysql rola pairs, in the format of

"<ldap_group>=<mysql_role>,<ldap_group2>=<mysql_role2>,..."

File: docs/security/simple-ldap-variables.md [hunk 4]
```diff
@@ -1,4 +1,28 @@
...
+| authentication_ldap_simple_group_search_attr | Name of the attribute that specifies the group names in the LDAP directory entries |
+| authentication_ldap_simple_group_search_filter | Custom group search filter |
+| authentication_ldap_simple_init_pool_size | Initial size of the connection pool to the LDAP server |
+| authentication_ldap_simple_log_status | logging level |
+| authentication_ldap_simple_max_pool_size | Maximum size of the pool of connections to the LDAP server |
+| authentication_ldap_simple_server_host | LDAP server host |
+| authentication_ldap_simple_server_port | LDAP server TCP/IP port number |
```
Comment (dutow) [inline docs/security/simple-ldap-variables.md]: The fallback variables aren't mentioned.

File: docs/security/simple-ldap-variables.md [hunk 5]
```diff
@@ -1,4 +1,28 @@
...
+| authentication_ldap_simple_ssl | If plugin connections to the LDAP server use the SSL protocol (ldaps://) |
+| authentication_ldap_simple_tls | If plugin connections to the LDAP server are secured with STARTTLS (ldap://) |
+| authentication_ldap_simple_user_search_attr | Name of the attribute that specifies user names in the LDAP directory entries |
+
+## Static variables
 
 The following variables are static and can only be changed at runtime.
```
Comment (dutow) [inline docs/security/simple-ldap-variables.md]: These variables are dynamic.

File: docs/security/simple-ldap.md (hunk 1 above)
Comment (patrickbirch) [reply docs/security/simple-ldap.md]: Fixed.

File: docs/security/simple-ldap.md (hunk 2 above)
Comment (patrickbirch) [reply docs/security/simple-ldap.md]: Removed.

File: docs/security/simple-ldap-variables.md (hunk 5 above)
Comment (patrickbirch) [reply docs/security/simple-ldap-variables.md]: Changed to Yes.

File: docs/security/simple-ldap-variables.md (hunk 3 above)
Comment (patrickbirch) [reply docs/security/simple-ldap-variables.md]: updated

File: docs/security/simple-ldap-variables.md (hunk 4 above)
Comment (patrickbirch) [reply docs/security/simple-ldap-variables.md]: Added

**Review by alina-derkach-oaza — APPROVED** [review]
LGTM

File: docs/security/sasl-based-ldap-variables.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+### `authentication_ldap_sasl_fallback_server_port`
+
+| Option | Description |
+|---|---|
+| Command-line | --authentication-ldap-sasl-fallback-server-port |
+| Scope | Global |
+| Dynamic | Yes |
+| Type | Sting |
+| Default | Null |
```
Comment (dutow) [inline docs/security/sasl-based-ldap-variables.md]: default is 0

File: docs/security/sasl-based-ldap-variables.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+If the primary server is unavailable, the authentication plugin attempts to connect to the fallback server and authenticate using that server.
+
+### `authentication_ldap_sasl_fallback_server_port`
+
+| Option | Description |
+|---|---|
+| Command-line | --authentication-ldap-sasl-fallback-server-port |
+| Scope | Global |
+| Dynamic | Yes |
+| Type | Sting |
```
Comment (dutow) [inline docs/security/sasl-based-ldap-variables.md]: integer

File: docs/security/sasl-based-ldap-variables.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+| Variable name | Description |
+|---|---|
+| authentication_ldap_sasl_bind_base_dn | Base distinguished name |
+| [authentication_ldap_sasl_bind_root_dn](#authentication_ldap_sasl_bind_root_dn) | Root distinguished name |
+| [authentication_ldap_sasl_bind_root_dn_pwd](#authentication_ldap_sasl_bind_root_pwd) | Password for the root distinguished name |
+| [authentication_ldap_sasl_ca_path](#authentication_ldap_sasl_ca_path) | Absolute path of the certificate authority |
+| [authentication_ldap_sasl_fallback_server_host](#authentication_ldap_sasl_fallback_server_host) | If the primary server is unavailable, the authentication plugin attempts to connect to the fallback server |
+| [authentication_ldap_sasl_fallback_server_port](#authentication_ldap_sasl_fallback_server_port) | The port number for the fallback server |
+| [authentication_ldap_sasl_group_role_mapping](#authentication_ldap_sasl_group_role_mapping) | LDAP group name |
```
Comment (dutow) [inline docs/security/sasl-based-ldap-variables.md]: not updated here - please make sure that the sasl and simple docs contain the same information.

File: docs/security/simple-ldap-variables.md [hunk 9]
```diff
@@ -1,100 +1,158 @@
-# Simple LDAP Variables
-
-The following variables are static and can only be changed at runtime.
```
Comment (dutow) [inline docs/security/simple-ldap-variables.md:3]: Sorry, my comment here was incorrect previously: some variables are static, just not all of them - previously everything was grouped under the "static variables" section. Generally the connection variables (host, port, ssl, ...) are static, but everything else (including most new variables) are dynamic.

File: docs/security/sasl-based-ldap-variables.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
+# SASL-based LDAP variables
+
+## Variables installed
+
+The installation adds the following variables:
+
+| Variable name | Description |
+|---|---|
+| authentication_ldap_sasl_bind_base_dn | Base distinguished name |
```
Comment (dutow) [inline docs/security/sasl-based-ldap-variables.md]: It seems that some variables have links, others don't have one. Is this intentional?

Comment (patrickbirch) [reply docs/security/sasl-based-ldap-variables.md]: Yes. I can only link to the variables with details. If you want to provide the details for the rest of the variables I'll add them.

File: docs/security/sasl-based-ldap-variables.md (hunk 8 above)
Comment (patrickbirch) [reply docs/security/sasl-based-ldap-variables.md]: Done.

File: docs/security/sasl-based-ldap-variables.md (hunk 7 above)
Comment (patrickbirch) [reply docs/security/sasl-based-ldap-variables.md]: Updated in SASL and simple

File: docs/security/ldap-authentication.md [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+* The `authentication_ldap_sasl` plugin, on the server, performs the SASL-based LDAP authentication. The client must use the `authentication_ldap_sasl_client` plugin. The method does not send the password to the server in cleartext. The server-side and client-side plugins use Simple Authentication and Security Layer (SASL) to send secure messages within the LDAP protocol.
+
+For either method, the database server rejects the connection if the client user name and the host name do not match a server account.
+
+If a database server LDAP authentication is successful, the LDAP server searches for an entry. The LDAP server matches the user and authenticates using the LDAP password. If the database server account names the LDAP user distinguished name (DN), added by the [`BY clause`](#use-the-by-clause-that-specifies-the-authentication-string), the LDAP server uses that value and the LDAP password provided by the client. This method fails in the following conditions:
+
+* DN and password have incorrect values
+* DN and password are empty and the LDAP server does not permit anonymous connections
```
Comment (dutow) [inline docs/security/ldap-authentication.md]: we do not allow anonymous connections (login with an empty password)

File: docs/security/ldap-authentication.md [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+=== "Use the authentication string in SASL-based LDAP"
+
+    If you provide the optional authentication string clause, ‘cn,ou,dc,dc’ in the example, the string is stored along with the password.
+
+    ```text
+    mysql> CREATE USER ... IDENTIFIED WITH authentication_ldap_sasl BY 'cn=[user name],ou=[organization unit],dc=[domain component],dc=com'
+    ```
+
+    The following variables are not used when creating a user with an authentication string:
```
Comment (dutow) [inline docs/security/ldap-authentication.md]: bind_root_dn and bind_root_pwd is also used by the new group_mapping variable.

File: docs/security/ldap-authentication.md [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    * authentication_ldap_sasl_group_search_attr
+
+    If the user is created with “IDENTIFIED BY authentication_ldap_sasl” the
+    listed variables are used.
+
+## Examples
+
+The following sections are examples of using simple LDAP authentication and SASL-based LDAP authentication.
+
+For the purposes of this example, we use the following OpenLDAP group:
```
Comment (dutow) [inline docs/security/ldap-authentication.md]: OpenLDAP user?

File: docs/security/sasl-based-ldap-variables.md (hunk 6 above)
Comment (patrickbirch) [reply docs/security/sasl-based-ldap-variables.md]: Fixed.

File: docs/security/simple-ldap-variables.md (hunk 9 above)
Comment (patrickbirch) [reply docs/security/simple-ldap-variables.md:3]: Checked against MySQL system variables.

File: docs/security/ldap-authentication.md (hunk 11 above)
Comment (patrickbirch) [reply docs/security/ldap-authentication.md]: rewritten.

File: docs/security/ldap-authentication.md (hunk 13 above)
Comment (patrickbirch) [reply docs/security/ldap-authentication.md]: Fixed.

File: docs/security/ldap-authentication.md (hunk 12 above)
Comment (patrickbirch) [reply docs/security/ldap-authentication.md]: Rewritten.

File: docs/security/ldap-authentication.md [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+
+=== "Use the authentication string in SASL-based LDAP"
+
+    If you provide the optional authentication string clause, ‘cn,ou,dc,dc’ in the example, the string is stored along with the password.
+
+    ```text
+    mysql> CREATE USER ... IDENTIFIED WITH authentication_ldap_sasl BY 'cn=[user name],ou=[organization unit],dc=[domain component],dc=com'
+    ```
+
+    Creating a user with an authentication string does not use the following system variables, unless the user is created with [authentication_ldap_sasl_group_role_mapping](ldap-system-variables.md#authentication_ldap_sasl_group_role_mapping):
```
Comment (dutow) [inline docs/security/ldap-authentication.md]: unless the authentication_ldap_sasl_group_role_mapping variable is used.

Comment (patrickbirch) [reply docs/security/ldap-authentication.md]: Updated.

**Review by dutow — APPROVED** [review]

**Pull request merged**